docs/source/providers/external.md

```diff
@@ -0,0 +1,175 @@
+# External Providers
+
+Llama Stack supports external providers that live outside of the main codebase. This allows you to:
+- Create and maintain your own providers independently
+- Share providers with others without contributing to the main codebase
+- Keep provider-specific code separate from the core Llama Stack code
+
+## Configuration
+
+To enable external providers, you need to configure the `external_providers_dir` in your Llama Stack configuration. This directory should contain your external provider specifications:
+
+```yaml
+external_providers_dir: /etc/llama-stack/providers.d/
+```
+
+## Directory Structure
+
+The external providers directory should follow this structure:
+
+```
+providers.d/
+  inference/
+    custom_ollama.yaml
+    vllm.yaml
+  vector_io/
+    qdrant.yaml
+  safety/
+    llama-guard.yaml
+```
+
+Each YAML file in these directories defines a provider specification for that particular API.
+
+## Provider Specification
+
+A provider specification is a YAML file that contains:
+- The adapter configuration
+- Required dependencies
+- API dependencies
+
+Here's an example for a custom Ollama provider:
+
+```yaml
+adapter:
+  adapter_type: custom_ollama
+  pip_packages: ["ollama", "aiohttp"]
+  config_class: llama_stack_ollama_provider.config.OllamaImplConfig
+  module: llama_stack_ollama_provider
+api_dependencies: []
+optional_api_dependencies: []
+```
+
+### Adapter Configuration
+
+The `adapter` section defines how to load and configure the provider:
+
+- `adapter_type`: A unique identifier for this adapter
+- `pip_packages`: List of Python packages required by the provider
+- `config_class`: The full path to the configuration class
+- `module`: The Python module containing the provider implementation
+
+### Dependencies
+
+- `api_dependencies`: List of Llama Stack APIs that this provider depends on
+- `optional_api_dependencies`: List of optional Llama Stack APIs that this provider can use
+
+## Required Implementation
+
+External providers must expose a `get_adapter_impl()` function in their module that takes two arguments:
+1. `config`: An instance of the provider's config class
+2. `deps`: A dictionary of API dependencies
```

Comment:
Since we are defining an external interface, perhaps we should consider if the names chosen are ideal here:

- what does "adapter" mean in `get_adapter_impl`? Should it be `get_provider_impl` perhaps?
- would it make sense to rename `deps` into `api_dependencies` to mimic the field name from YAML?

Reply:
++ @booxter

Reply:
I think we can erase the adapter vs provider distinction. It hasn't proven useful.